### PR TITLE
Prioritize "Before Class End" of Blank Lines Option #4611

### PR DESCRIPTION
**Pull request by junichi11**
- https://github.com/apache/netbeans/issues/4611
- Fix unit tests

#### Example
```php
<?php

class GH4611Class
{

private int $field;

public function method1() {
}

/**
 * Test.
 */
public function method2() {
}

}
```
##### Formatting Options:

nb-php-gh-4611-blank-lines-options

nb-php-gh-4611-braces-options

##### Before
```php
<?php

class GH4611Class
{
    private int $field;

    public function method1() {
        
    }

    /**
     * Test.
     */
    public function method2() {
        
    }

}
```

##### After
```php
<?php

class GH4611Class
{
    private int $field;

    public function method1() {
        
    }

    /**
     * Test.
     */
    public function method2() {
        
    }
}
```

### CODE REVIEW

File: php/php.editor/src/org/netbeans/modules/php/editor/indent/FormatVisitor.java [hunk 1]
```diff
@@ -600,7 +600,7 @@ public void visit(Block node) {
                 formatTokens.add(new FormatToken.IndentToken(ts.offset(), options.indentSize));
             }
 
-            if (parent instanceof ClassDeclaration || parent instanceof InterfaceDeclaration || parent instanceof TraitDeclaration) {
+            if (isTypeNode(parent)) {
```
Comment (junichi11) [inline php/php.editor/src/org/netbeans/modules/php/editor/indent/FormatVisitor.java:603]: Fix missing `EnumDeclaration`

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/indent/PHPFormatterWrappingTest.java [hunk 2]
```diff
@@ -686,6 +686,12 @@ public void testCoalescingOp03a() throws Exception {
         reformatFileContents("testfiles/formatting/wrapping/coalescingOp03a.php", options);
     }
 
+    public void testCoalescingOp03b() throws Exception {
```
Comment (junichi11) [inline php/php.editor/test/unit/src/org/netbeans/modules/php/editor/indent/PHPFormatterWrappingTest.java:689]: Maybe, I overwrote this method before I copy this method (`testCoalescingOp03b()`->`testCoalescingOp04a()`)

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/indent/PHPFormatterTestBase.java [hunk 3]
```diff
@@ -119,10 +119,14 @@ protected void reformatFileContents(String file, Map<String, Object> options) th
...
 
     protected void reformatFileContents(String file, Map<String, Object> options, boolean isTemplate) throws Exception {
-        reformatFileContents(file, options, isTemplate, new IndentPrefs(4, 4));
+        reformatFileContents(file, options, isTemplate, false, new IndentPrefs(4, 4));
     }
 
-    private void reformatFileContents(String file, Map<String, Object> options, boolean isTemplate, IndentPrefs indentPrefs) throws Exception {
+    protected void reformatFileContents(String file, Map<String, Object> options, boolean isTemplate, boolean includeTestName) throws Exception {
+        reformatFileContents(file, options, isTemplate, includeTestName, new IndentPrefs(4, 4));
+    }
```
Comment (junichi11) [inline php/php.editor/test/unit/src/org/netbeans/modules/php/editor/indent/PHPFormatterTestBase.java:127]: For using the same test file. (Avoid putting the same file for each test.)

**Review by tmysik — APPROVED** [review]
Thanks for your work!

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/indent/PHPFormatterWrappingTest.java (hunk 2 above)
Comment (tmysik) [reply php/php.editor/test/unit/src/org/netbeans/modules/php/editor/indent/PHPFormatterWrappingTest.java:689]: Not sure what you mean exactly, sorry...

**Comment by tmysik** [comment]
BTW, if we would like to avoid having so many changed test files, we could change their formatting options (blank lines before class end: 1). But no problem, of course. Just an idea.

Will wait for your feedback (here and above), then I am merging. Thanks!

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/indent/PHPFormatterWrappingTest.java (hunk 2 above)
Comment (junichi11) [reply php/php.editor/test/unit/src/org/netbeans/modules/php/editor/indent/PHPFormatterWrappingTest.java:689]: When I added the next unit test(`testCoalescingOp04a()`), I've removed `testCoalescingOp03b()` accidentally.  It seems that I didn't notice that and I've committed and pushed it...

**Comment by junichi11** [comment]
Will change the default value later :) Thank you, Tomas!

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/indent/PHPFormatterWrappingTest.java (hunk 2 above)
Comment (tmysik) [reply php/php.editor/test/unit/src/org/netbeans/modules/php/editor/indent/PHPFormatterWrappingTest.java:689]: I see now, thanks!

File: php/php.editor/src/org/netbeans/modules/php/editor/indent/FmtOptions.java [hunk 4]
```diff
@@ -288,7 +288,7 @@ private static void createDefaults() {
             {BLANK_LINES_BEFORE_CLASS, "1"}, //NOI18N
             {BLANK_LINES_AFTER_CLASS, "1"}, //NOI18N
             {BLANK_LINES_AFTER_CLASS_HEADER, "0"}, //NOI18N
-            {BLANK_LINES_BEFORE_CLASS_END, "0"}, //NOI18N
+            {BLANK_LINES_BEFORE_CLASS_END, "1"}, //NOI18N
```
[the remark this answers is not in the document]
Comment (tmysik) [reply php/php.editor/src/org/netbeans/modules/php/editor/indent/FmtOptions.java:291]: I am sorry for not being clear enough in my first comment

[the remark this answers is not in the document]
Comment (junichi11) [reply php/php.editor/src/org/netbeans/modules/php/editor/indent/FmtOptions.java:291]: Ah, I see now.

I'll revert this PR to my first commit :)

> I am sorry for not being clear enough in my first comment

No problem :)

#### Example
```php
class GH4611Class {

    private int $field;

    public function method1() {
        
    }

}
```


#### CLASS_END: 0
Existing behavior is changed by default options

Before PR by default options

```php
class GH4611Class {

    private int $field;

    public function method1() {
        
    }

}
```

After PR by default options

```php
class GH4611Class {

    private int $field;

    public function method1() {
        
    }
}
```

#### CLASS_END: 1
Existing behavior is keeped by default options

Before PR by default options

```php
class GH4611Class {

    private int $field;

    public function method1() {
        
    }

}
```

After PR by default options

```php
class GH4611Class {

    private int $field;

    public function method1() {
        
    }

}
```

[the remark this answers is not in the document]
Comment (tmysik) [reply php/php.editor/src/org/netbeans/modules/php/editor/indent/FmtOptions.java:291]: I see now, it looks OK to me. Thanks!

**Comment by junichi11** [comment]
OK. Maybe, I'll push it tomorrow. Thanks!

**Comment by junichi11** [comment]
@tmysik I've reverted it. If it's OK for you, let's merge it. Thank you!

**Comment by zeleznypa** [comment]
❤️

**Pull request merged**